### PR TITLE
Improve Operator upgrade mechanism

### PR DESCRIPTION
**Pull request by nikhil-thomas**
Improve ReadyChecks for Pipelines and Triggers

Add mechanism to detect upgrade scenario in reconcilers and update
status with relevant messages

Minor fixes to ensure TektonAddon reconciler acts more predictably

Signed-off-by: Nikhil Thomas <nikthoma@redhat.com>

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by nikhil-thomas** [comment]
@concaf @vdemeester 
ready for review

File: pkg/reconciler/common/common.go [hunk 1]
```diff
@@ -77,3 +93,19 @@ func getTriggerRes(informer informer.TektonTriggerInformer) (*v1alpha1.TektonTri
...
 	return res, err
 }
+
+func CheckUpgradePending(tc v1alpha1.TektonComponent) (bool, error) {
+	labels := tc.GetLabels()
+	ver, ok := labels[tektoninstallerset.ReleaseVersionKey]
+	if !ok {
+		return true, nil
+	}
+	operatorVersion, err := OperatorVersion(context.TODO())
```
Comment (vdemeester) [inline pkg/reconciler/common/common.go:103]: nit, for the future: we should probably "chain" `context` 👼🏼

**Review by vdemeester — APPROVED** [review]

File: pkg/reconciler/kubernetes/tektontrigger/tektontrigger.go [hunk 2]
```diff
@@ -146,6 +146,11 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, tt *v1alpha1.TektonTrigg
 
 	// Pass the object through defaulting
 	tt.SetDefaults(ctx)
+	// Mark TektonPipeline Instance as Not Ready if an upgrade is needed
```
Comment (piyush-garg) [inline pkg/reconciler/kubernetes/tektontrigger/tektontrigger.go:149]: nit: TektonTrigger

File: pkg/reconciler/openshift/tektonaddon/tektonaddon.go [hunk 3]
```diff
@@ -121,6 +121,15 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, ta *v1alpha1.TektonAddon
 		return nil
 	}
 
+	// Pass the object through defaulting
+	ta.SetDefaults(ctx)
+
+	// Mark TektonPipeline Instance as Not Ready if an upgrade is needed
```
Comment (piyush-garg) [inline pkg/reconciler/openshift/tektonaddon/tektonaddon.go:127]: nit: TektonAddon

File: pkg/reconciler/shared/tektonconfig/tektonconfig.go [hunk 4]
```diff
@@ -94,6 +96,11 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, tc *v1alpha1.TektonConfi
 	}
 
 	tc.SetDefaults(ctx)
+	// Mark TektonPipeline Instance as Not Ready if an upgrade is needed
```
Comment (piyush-garg) [inline pkg/reconciler/shared/tektonconfig/tektonconfig.go:99]: nit: TektonConfig

File: pkg/reconciler/kubernetes/tektonpipeline/tektonpipeline.go [hunk 5]
```diff
@@ -411,3 +417,26 @@ func (m *Recorder) logMetrics(status, version string, logger *zap.SugaredLogger)
...
+		tp.Status.MarkInstallerSetNotReady("Upgrade Pending")
+		tp.Status.MarkNotReady("Upgrade Pending")
+	}
+	if labels == nil {
+		labels = map[string]string{}
+	}
+	labels[tektoninstallerset.ReleaseVersionKey] = r.operatorVersion
+	tp.SetLabels(labels)
+
+	if _, err := r.operatorClientSet.OperatorV1alpha1().TektonPipelines().Update(ctx,
```
Comment (sm43) [inline pkg/reconciler/kubernetes/tektonpipeline/tektonpipeline.go:437]: in future, we should try to convey this through status 
may be `status.phase` and try to minimize updating the "object" in reconciler
..
many times it reconciles bcz of conflict error while updating
...

...

**Comment by sm43** [comment]
```
=========================================
==== TESTING IF GOLINT HAS BEEN DONE ====
=========================================
pkg/reconciler/kubernetes/tektoninstallerset/const.go:1: : import cycle not allowed in test (typecheck)
/*
==============================
==== GO LINT TESTS FAILED ====
==============================
```
lint check seems to be failing

LGTM !

File: pkg/reconciler/kubernetes/tektonpipeline/tektonpipeline.go [hunk 6]
```diff
@@ -131,6 +131,12 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, tp *v1alpha1.TektonPipel
 	// Pass the object through defaulting
 	tp.SetDefaults(ctx)
 
+	// Mark TektonPipeline Instance as Not Ready if an upgrade is needed
+	err := r.markUpgrade(ctx, tp)
+	if err != nil {
+		return err
+	}
```
Comment (savitaashture) [inline pkg/reconciler/kubernetes/tektonpipeline/tektonpipeline.go:138]:
```suggestion
	if err := r.markUpgrade(ctx, tp); err != nil {
		return err
	}
```

File: pkg/reconciler/kubernetes/tektonpipeline/tektonpipeline.go [hunk 7]
```diff
@@ -411,3 +417,26 @@ func (m *Recorder) logMetrics(status, version string, logger *zap.SugaredLogger)
...
+func (r *Reconciler) markUpgrade(ctx context.Context, tp *v1alpha1.TektonPipeline) error {
+	labels := tp.GetLabels()
+	ver, ok := labels[tektoninstallerset.ReleaseVersionKey]
+	if ok && ver == r.operatorVersion {
+		return nil
+	}
+	if ok && ver != r.operatorVersion {
+		tp.Status.MarkInstallerSetNotReady("Upgrade Pending")
+		tp.Status.MarkNotReady("Upgrade Pending")
+	}
```
Comment (savitaashture) [inline pkg/reconciler/kubernetes/tektonpipeline/tektonpipeline.go:430]:
```suggestion
	if ok {
	    if ver == r.operatorVersion {
		return nil
	    }
	    tp.Status.MarkInstallerSetNotReady("Upgrade Pending")
	    tp.Status.MarkNotReady("Upgrade Pending")
	}
```

File: pkg/reconciler/kubernetes/tektonpipeline/tektonpipeline.go (hunk 5 above)
Comment (nikhil-thomas) [reply pkg/reconciler/kubernetes/tektonpipeline/tektonpipeline.go:437]: agreed :+1: 

this time i was just avoiding adding a new item to all statuses (operatorVersion).

File: pkg/reconciler/kubernetes/tektonpipeline/tektonpipeline.go (hunk 7 above)
Comment (nikhil-thomas) [reply pkg/reconciler/kubernetes/tektonpipeline/tektonpipeline.go:430]: i shall make this update. :+1: 
however, i find nested `if` less effective in conveying the logic to the reader generally. :technologist:. ie, reader's mind has to backtrack.

File: pkg/reconciler/common/common.go (hunk 1 above)
Comment (nikhil-thomas) [reply pkg/reconciler/common/common.go:103]: absolutely, :+1:

**Review by pradeepitm12 — COMMENTED** [review]
/lgtm

**Pull request merged**